Use smaller data type to represent the counter fields of FunctionBody.

leirocks opened the pull request:
Use smaller data type to represent the counter fields of FunctionBody.

Most of the counters has value less than 256, use single byte to hold them. And can promote to short or full 32 bit integer when necessary.

ianwjhalliday commented on lib/Runtime/Base/FunctionBody.h:
```diff
@@ -1658,6 +1660,74 @@ namespace Js
+
+                // Signed integers need keep the sign when promoting 
+                SignedFieldsStart                       = 26,
+                SerializationIndex                      = 26,
```
These two are both 26

leirocks replied:
SignedFieldsStart  is a marker that from that point, it saves signed value. maybe I can remove this, instead, use a byte in the structure to save the position -- there're 3 bytes at the beginning is not used.

ianwjhalliday replied:
Oh gotcha, I missed that SignedFieldsStart was a label marking the beginning of special fields.

ianwjhalliday commented:
With this and your previous change how much does this shrink FunctionBody down on x64?

ianwjhalliday commented:
Ignoring the separately added data for the pointers and counters.  I'm curious about sizeof(FunctionBody) for curiosity's sake.

leirocks commented:
@ianwjhalliday it reduced from 648 to 384 bytes

leirocks commented:
@dotnet-bot test this please

ianwjhalliday commented:
Nice, that is a significant reduction.

leirocks commented:
@dotnet-bot test EOL Check please

curtisman commented on lib/Runtime/Library/ScriptFunction.cpp:
```diff
@@ -594,7 +594,7 @@ namespace Js
                         memset(this->m_inlineCaches[i], 0, sizeof(InlineCache));
                     }
-                    else if(!scriptContext->IsClosed())
+                    else if(scriptContext != nullptr && !scriptContext->IsClosed())
```
In what situation can the scriptContext be nullptr?

leirocks replied:
ah, this is not intentional change. previously in order to solve the issue the when FunctionBody::Cleanup sometimes the scriptContext is already gone, I set this to null. So this is not needed any more because when cleaning up we should not be expanding the structure so should not need the recycler from scriptContext anymore.
let me remove the change. thanks.

curtisman commented on lib/Runtime/Base/CompactCounters.h:
```diff
@@ -0,0 +1,269 @@
+namespace Js
+{
+    template<class T, typename CountT = T::CounterFields>
+    struct Counter
```
This is name is kind of generic?  Rename to CompactCounters?

curtisman commented on lib/Runtime/Base/CompactCounters.h:
```diff
@@ -0,0 +1,269 @@
+            else 
+            {
+                Assert(!ThreadContext::GetContextForCurrentThread());
+                AutoCriticalSection autocs(T::GetLock());
```
Is there any interpreter perf impact from this lock?  Is it rare enough to be have to use a larger size that it doesn't matter?

curtisman commented on lib/Runtime/Base/CompactCounters.h:
```diff
@@ -0,0 +1,269 @@
+            return this->i32Fileds[type] = val;
+        }
+
+        uint32 Increase(CountT typeEnum, T* host)
```
Increment?

curtisman commented on lib/Runtime/Base/CompactCounters.h:
```diff
@@ -0,0 +1,269 @@
+        union {
+            uint8 u8fields[CountT::Max];
+            int8 i8fields[CountT::Max];
+            WriteBarrierPtr<uint16> u16Fileds;
```
Fileds = Fields?

curtisman commented on lib/Runtime/Base/CompactCounters.h:
```diff
@@ -0,0 +1,269 @@
+            }
+            else if (localFieldSize == 2)
+            {
+                value = this->u16Fileds[type];
```
Since we didn't take a lock here, background thread could potentially read old value from u8fields here even after we switch to u16fields.   May guess it that it is probably ok, because we didn't have a lock before, so the background thread either is just access field that doesn't change after byte code gen, or some count that doesn't matter if we missed some.     But please verify that is true.

Please add a comment here too, so we won't think it is a bug in the future.

leirocks commented:
@curtisman as discussed, moved following mutable fields out of the structure. and added Assert in case there's any other fields missed.

        uint32 interpretedCount;
        uint32 loopInterpreterLimit;
        uint32 debuggerScopeIndex;
        uint32 savedPolymorphicCacheState;

curtisman commented:
LGTM
